## Goal:
Take in the famous MNIST data set of handwritten digits and create a Deep Learning model using TensforFlow to correctly classify them.

## Steps:
I will take the following steps to acheive my goal:
1. Set model parameters
2. Set placeholders, weights, and biases
3. Define perceptron, cost, and optimizer
4. Run the model
5. Test the model
6. Refine the model

In [1]:
import tensorflow as tf

In [2]:
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets("/tmp/data/",one_hot=True)

Extracting /tmp/data/train-images-idx3-ubyte.gz
Extracting /tmp/data/train-labels-idx1-ubyte.gz
Extracting /tmp/data/t10k-images-idx3-ubyte.gz
Extracting /tmp/data/t10k-labels-idx1-ubyte.gz


## 1.0 Setting the parameters
Here are the initial paramters that I am setting for the first run through this model.

In [3]:
learning_rate = 0.001
training_epochs = 15
batch_size = 100

In [4]:
n_hidden_1 = 256
n_hidden_2 = 256
n_input = 784
n_classes = 10
n_samples = mnist.train.num_examples

## 2.0 Placeholders, weights, and biases
These are the weights and biases that will be applied to the model as the input travels through the hidden layers and to the output layer.

In [5]:
x = tf.placeholder('float',[None,n_input])
y = tf.placeholder('float',[None,n_classes])

In [6]:
weights = {
    'h1': tf.Variable(tf.random_normal([n_input,n_hidden_1])),
    'h2': tf.Variable(tf.random_normal([n_hidden_1,n_hidden_2])),
    'out': tf.Variable(tf.random_normal([n_hidden_2,n_classes])),
}

biases = {
    'b1': tf.Variable(tf.random_normal([n_hidden_1])),
    'b2': tf.Variable(tf.random_normal([n_hidden_2])),
    'out': tf.Variable(tf.random_normal([n_classes])),
}

## 3.0 Define perceptron, cost, and optimizer
I have chosen to use two hidden layers in this deep learning model. Adding more hidden layers would increase the accuracy of the model, however with this dataset 2 hidden layers is enough to achieve a good accuracy. 

My activation fucntion will be RELU and the optimizer will be the Adam Optimizer.

In [7]:
def multilayer_perceptron(x,weights,biases):
    
    layer_1 = tf.add(tf.matmul(x,weights['h1']),biases['b1'])
    layer_1 = tf.nn.relu(layer_1)
    
    layer_2 = tf.add(tf.matmul(layer_1,weights['h2']),biases['b2'])
    layer_2 = tf.nn.relu(layer_2)
    
    out_layer = tf.matmul(layer_2,weights['out']) + biases['out']
    return out_layer

pred = multilayer_perceptron(x,weights,biases)

In [8]:
# Cost and optimiser 

cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=pred,labels=y))
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost)

## 4.0 Running the session
Now it's time to run our model. The two for loops will loop through each epoch and batch respectively and then print out the cost for each epoch.

In [11]:
init = tf.global_variables_initializer()
sess = tf.InteractiveSession()
sess.run(init)

for epoch in range(training_epochs):

    avg_cost = 0.0
    total_batch = int(n_samples/batch_size)

    for i in range(total_batch):

        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, c = sess.run([optimizer, cost], feed_dict={x: batch_x, y: batch_y})
        avg_cost += c / total_batch

    print("Epoch: {} cost={:.4f}".format(epoch+1,avg_cost))

print("Model has completed {} Epochs of Training".format(training_epochs))

Epoch: 1 cost=165.4091
Epoch: 2 cost=42.0048
Epoch: 3 cost=27.1312
Epoch: 4 cost=19.0818
Epoch: 5 cost=14.2210
Epoch: 6 cost=10.6880
Epoch: 7 cost=8.0346
Epoch: 8 cost=6.0636
Epoch: 9 cost=4.4497
Epoch: 10 cost=3.4474
Epoch: 11 cost=2.5600
Epoch: 12 cost=1.9112
Epoch: 13 cost=1.3684
Epoch: 14 cost=1.2052
Epoch: 15 cost=0.8536
Model has completed 15 Epochs of Training


## 5.0 Testing the model
Now that I have ran through the first iteration of our model, it is time to evalulate the accuracy of what I created.

In [15]:
correct_predictions = tf.equal(tf.argmax(pred, 1), tf.argmax(y, 1))
correct_predictions = tf.cast(correct_predictions, "float")
accuracy = tf.reduce_mean(correct_predictions)
print("Accuracy:", accuracy.eval({x: mnist.test.images, y: mnist.test.labels}))

Accuracy: 0.9448


## 6.0 Improving the model
Now it is time to improve our model by tweaking the parameters. The parameters that I set at the beginning are what I will be adjusting. 

Deep learning in a fast moving field, and the learning rate finder has been a topic of discussion recently. The idea being that you can create a model that gradually increases the learning rate to find the optimum for a number of batch sizes. 

Here however I will not be using this model and using an intuition based approach where I assume that decreasing the batch size and decrasing the learning rate will reduce my loss and increase accuracy. 

In [21]:
# New parameters
learning_rate = 0.0001
training_epochs = 15
batch_size = 50

# Run the model again

init = tf.global_variables_initializer()
sess = tf.InteractiveSession()
sess.run(init)

for epoch in range(training_epochs):

    avg_cost = 0.0
    total_batch = int(n_samples/batch_size)

    for i in range(total_batch):

        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, c = sess.run([optimizer, cost], feed_dict={x: batch_x, y: batch_y})
        avg_cost += c / total_batch

    print("Epoch: {} cost={:.4f}".format(epoch+1,avg_cost))

print("Model has completed {} Epochs of Training".format(training_epochs))

correct_predictions = tf.equal(tf.argmax(pred, 1), tf.argmax(y, 1))
correct_predictions = tf.cast(correct_predictions, "float")
accuracy = tf.reduce_mean(correct_predictions)
print("Accuracy:", accuracy.eval({x: mnist.test.images, y: mnist.test.labels}))

Epoch: 1 cost=118.0878
Epoch: 2 cost=29.2770
Epoch: 3 cost=17.2393
Epoch: 4 cost=11.1376
Epoch: 5 cost=7.4225
Epoch: 6 cost=5.1362
Epoch: 7 cost=3.5494
Epoch: 8 cost=2.5115
Epoch: 9 cost=1.7533
Epoch: 10 cost=1.3747
Epoch: 11 cost=1.0053
Epoch: 12 cost=0.9000
Epoch: 13 cost=0.7001
Epoch: 14 cost=0.7300
Epoch: 15 cost=0.5834
Model has completed 15 Epochs of Training
Accuracy: 0.9556


By increasing my epoch rate I can also increase the accuracy

In [22]:
# New parameters
learning_rate = 0.0001
training_epochs = 25
batch_size = 50

# Run the model again

init = tf.global_variables_initializer()
sess = tf.InteractiveSession()
sess.run(init)

for epoch in range(training_epochs):

    avg_cost = 0.0
    total_batch = int(n_samples/batch_size)

    for i in range(total_batch):

        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, c = sess.run([optimizer, cost], feed_dict={x: batch_x, y: batch_y})
        avg_cost += c / total_batch

    print("Epoch: {} cost={:.4f}".format(epoch+1,avg_cost))

print("Model has completed {} Epochs of Training".format(training_epochs))

correct_predictions = tf.equal(tf.argmax(pred, 1), tf.argmax(y, 1))
correct_predictions = tf.cast(correct_predictions, "float")
accuracy = tf.reduce_mean(correct_predictions)
print("Accuracy:", accuracy.eval({x: mnist.test.images, y: mnist.test.labels}))

Epoch: 1 cost=119.1054
Epoch: 2 cost=31.7345
Epoch: 3 cost=18.4623
Epoch: 4 cost=12.0373
Epoch: 5 cost=7.9507
Epoch: 6 cost=5.6495
Epoch: 7 cost=3.8142
Epoch: 8 cost=2.7813
Epoch: 9 cost=2.0712
Epoch: 10 cost=1.5027
Epoch: 11 cost=1.1495
Epoch: 12 cost=0.9879
Epoch: 13 cost=0.9631
Epoch: 14 cost=0.8315
Epoch: 15 cost=0.5799
Epoch: 16 cost=0.6213
Epoch: 17 cost=0.5586
Epoch: 18 cost=0.5539
Epoch: 19 cost=0.4693
Epoch: 20 cost=0.4628
Epoch: 21 cost=0.4194
Epoch: 22 cost=0.4373
Epoch: 23 cost=0.3462
Epoch: 24 cost=0.4385
Epoch: 25 cost=0.3414
Model has completed 25 Epochs of Training
Accuracy: 0.9607


## Results
Through some tweaking of the initial parameters I was able to increase accuracy from 94% to 96%. The way I did this was:
- Decreasing the learning rate and batch size
- Increasing the epoch count